pxtlib/cpp.ts

```diff
@@ -732,33 +742,71 @@ namespace pxt.hex {
 
     function downloadHexInfoCoreAsync(extInfo: pxtc.ExtensionInfo) {
         let hexurl = ""
-        return getCdnUrlAsync()
-            .then(url => {
-                hexurl = url + "/compile/" + extInfo.sha
-                return U.httpGetTextAsync(hexurl + ".hex")
+
+        return downloadHexInfoLocalAsync(extInfo)
+            .then((hex) => {
+                if (hex) {
+                    // Found the hex image in the local server cache, use that
+                    return hex;
+                }
+
+                return getCdnUrlAsync()
+                    .then(url => {
+                        hexurl = url + "/compile/" + extInfo.sha
+                        return U.httpGetTextAsync(hexurl + ".hex")
+                    })
+                    .then(r => r, e =>
+                        Cloud.privatePostAsync("compile/extension", { data: extInfo.compileData })
+                            .then(ret => new Promise<string>((resolve, reject) => {
+                                let tryGet = () => Util.httpGetJsonAsync(ret.hex.replace(/\.hex/, ".json"))
+                                    .then(json => {
+                                        if (!json.success)
+                                            U.userError(JSON.stringify(json, null, 1))
+                                        else
+                                            resolve(U.httpGetTextAsync(hexurl + ".hex"))
+                                    },
+                                    e => {
+                                        setTimeout(tryGet, 1000)
+                                        return null
+                                    })
+                                tryGet();
+                            })))
+                    .then(text =>
+                        Util.httpGetJsonAsync(hexurl + "-metainfo.json")
```

Comment:
Removed the extra query, do we still need to return an object that has `enums` and `arrays` properties?